# HG changeset patch -- Bitbucket.org
# Project gevent
# URL http://bitbucket.org/denis/gevent/overview
# User Denis Bilenko <denis.bilenko@gmail.com>
# Date 1275064878 -25200
# Node ID aca8c6c649a6e5f15b01f0aa977a654956f78110
# Parent  bbe9526c07991b2c598eb3ce84985502e682fa96
Hub: store/clear/restore current exc_info before/after switching

--- a/gevent/hub.py
+++ b/gevent/hub.py
@@ -144,13 +144,18 @@ class Hub(greenlet):
     def switch(self):
         cur = getcurrent()
         assert cur is not self, 'Cannot switch to MAINLOOP from MAINLOOP'
-        switch_out = getattr(cur, 'switch_out', None)
-        if switch_out is not None:
-            try:
-                switch_out()
-            except:
-                traceback.print_exc()
-        return greenlet.switch(self)
+        exc_info = sys.exc_info()
+        try:
+            sys.exc_clear()
+            switch_out = getattr(cur, 'switch_out', None)
+            if switch_out is not None:
+                try:
+                    switch_out()
+                except:
+                    traceback.print_exc()
+            return greenlet.switch(self)
+        finally:
+            core.set_exc_info(*exc_info)
 
     def run(self):
         global _threadlocal
